File: scripts/transaction/test_session_pool.py
"""
P21 Transaction API - Session Pool Behavior Test

This script tests for session pool contamination issues by:
1. Making rapid successive requests
2. Logging success/failure patterns
3. Testing with various delays
4. Capturing session-related headers

Results help diagnose intermittent failures caused by dirty session pools.
"""

import asyncio
import httpx
import json
import os
import time
import random
from datetime import datetime
from pathlib import Path
from dataclasses import dataclass, field
from typing import Optional

from dotenv import load_dotenv
load_dotenv(Path(__file__).parent.parent / ".env")

BASE_URL = os.getenv("P21_BASE_URL", "https://play.p21server.com")
USERNAME = os.getenv("P21_USERNAME")
PASSWORD = os.getenv("P21_PASSWORD")

import warnings
warnings.filterwarnings("ignore")


@dataclass
class TestResult:
    """Result of a single API call."""
    attempt: int
    timestamp: str
    elapsed_ms: int
    success: bool
    status_code: int
    error_type: Optional[str] = None
    error_message: Optional[str] = None
    session_headers: dict = field(default_factory=dict)
    response_preview: str = ""


class SessionPoolTester:
    """Tests P21 Transaction API session pool behavior."""

    def __init__(self, base_url: str, username: str, password: str):
        self.base_url = base_url
        self.username = username
        self.password = password
        self.token = None
        self.ui_server_url = None
        self.results: list[TestResult] = []

    async def authenticate(self, client: httpx.AsyncClient) -> None:
        """Get authentication token."""
        resp = await client.post(
            f"{self.base_url}/api/security/token",
            headers={
                "username": self.username,
                "password": self.password,
                "Content-Type": "application/json",
                "Accept": "application/json"
            },
            content="",
            follow_redirects=True
        )
        resp.raise_for_status()
        self.token = resp.json()["AccessToken"]

        # Get UI server URL
        resp = await client.get(
            f"{self.base_url}/api/ui/router/v1?urlType=external",
            headers={"Authorization": f"Bearer {self.token}", "Accept": "application/json"},
            follow_redirects=True
        )
        resp.raise_for_status()
        self.ui_server_url = resp.json()["Url"].rstrip("/")

    def build_test_payload(self) -> dict:
        """Build a simple test payload that should succeed.

        Using SalesPricePage create since we know it works.
        """
        return {
            "Name": "SalesPricePage",
            "UseCodeValues": False,
            "Transactions": [
                {
                    "DataElements": [
                        {
                            "Name": "FORM.form",
                            "Type": "Form",
                            "Keys": [],
                            "Rows": [{
                                "Edits": [
                                    {"Name": "price_page_type_cd", "Value": "Supplier / Product Group"},
                                    {"Name": "company_id", "Value": "ACME"},
                                    {"Name": "supplier_id", "Value": 10.0},
                                    {"Name": "product_group_id", "Value": "FA5"},
                                    {"Name": "description", "Value": f"SESSION-TEST-{datetime.now().strftime('%H%M%S%f')}"},
                                    {"Name": "pricing_method_cd", "Value": "Source"},
                                    {"Name": "source_price_cd", "Value": "Supplier List Price"},
                                    {"Name": "effective_date", "Value": "2025-01-01"},
                                    {"Name": "expiration_date", "Value": "2030-12-31"},
                                    {"Name": "totaling_method_cd", "Value": "Item"},
                                    {"Name": "totaling_basis_cd", "Value": "Supplier List Price"},
                                    {"Name": "row_status_flag", "Value": "Active"}
                                ],
                                "RelativeDateEdits": []
                            }]
                        },
                        {
                            "Name": "VALUES.values",
                            "Type": "Form",
                            "Keys": [],
                            "Rows": [{
                                "Edits": [
                                    {"Name": "calculation_method_cd", "Value": "Multiplier"},
                                    {"Name": "calculation_value1", "Value": "0.5"}
                                ],
                                "RelativeDateEdits": []
                            }]
                        }
                    ],
                    "Status": "New"
                }
            ]
        }

    async def make_request(self, client: httpx.AsyncClient, attempt: int) -> TestResult:
        """Make a single Transaction API request and capture results."""
        timestamp = datetime.now().isoformat()
        start = time.perf_counter()

        payload = self.build_test_payload()

        try:
            resp = await client.post(
                f"{self.ui_server_url}/api/v2/transaction",
                headers={
                    "Authorization": f"Bearer {self.token}",
                    "Content-Type": "application/json",
                    "Accept": "application/json"
                },
                json=payload,
                follow_redirects=True,
                timeout=30.0
            )

            elapsed_ms = int((time.perf_counter() - start) * 1000)

            # Capture session-related headers
            session_headers = {
                k: v for k, v in resp.headers.items()
                if any(x in k.lower() for x in ['session', 'x-p21', 'server', 'instance'])
            }

            if resp.status_code == 200:
                data = resp.json()
                summary = data.get("Summary", {})
                succeeded = summary.get("Succeeded", 0)
                failed = summary.get("Failed", 0)
                messages = data.get("Messages", [])

                if succeeded > 0 and failed == 0:
                    return TestResult(
                        attempt=attempt,
                        timestamp=timestamp,
                        elapsed_ms=elapsed_ms,
                        success=True,
                        status_code=200,
                        session_headers=session_headers,
                        response_preview=f"Succeeded: {succeeded}"
                    )
                else:
                    error_msg = messages[0] if messages else "Unknown error"
                    return TestResult(
                        attempt=attempt,
                        timestamp=timestamp,
                        elapsed_ms=elapsed_ms,
                        success=False,
                        status_code=200,
                        error_type="ValidationError",
                        error_message=str(error_msg)[:200],
                        session_headers=session_headers,
                        response_preview=f"Failed: {failed}, Messages: {len(messages)}"
                    )
            else:
                error_text = resp.text[:500]
                error_type = "UnexpectedWindow" if "Unexpected response window" in error_text else "HTTPError"

                return TestResult(
                    attempt=attempt,
                    timestamp=timestamp,
                    elapsed_ms=elapsed_ms,
                    success=False,
                    status_code=resp.status_code,
                    error_type=error_type,
                    error_message=error_text[:200],
                    session_headers=session_headers,
                    response_preview=error_text[:100]
                )

        except Exception as e:
            elapsed_ms = int((time.perf_counter() - start) * 1000)
            return TestResult(
                attempt=attempt,
                timestamp=timestamp,
                elapsed_ms=elapsed_ms,
                success=False,
                status_code=0,
                error_type=type(e).__name__,
                error_message=str(e)[:200],
                response_preview=""
            )

    async def run_rapid_test(self, count: int = 10, delay_ms: int = 0) -> list[TestResult]:
        """Run rapid successive requests with optional delay."""
        results = []

        async with httpx.AsyncClient(verify=False, timeout=60.0) as client:
            await self.authenticate(client)

            for i in range(count):
                result = await self.make_request(client, i + 1)
                results.append(result)

                status = "OK" if result.success else "FAIL"
                print(f"  [{i+1:2d}] {status} {result.elapsed_ms:4d}ms - {result.response_preview[:50]}")

                if delay_ms > 0 and i < count - 1:
                    await asyncio.sleep(delay_ms / 1000)

        return results

    async def run_parallel_test(self, count: int = 5) -> list[TestResult]:
        """Run parallel requests to stress test session pool."""
        async with httpx.AsyncClient(verify=False, timeout=60.0) as client:
            await self.authenticate(client)

            tasks = [self.make_request(client, i + 1) for i in range(count)]
            results = await asyncio.gather(*tasks)

            for result in results:
                status = "OK" if result.success else "FAIL"
                print(f"  [{result.attempt:2d}] {status} {result.elapsed_ms:4d}ms - {result.response_preview[:50]}")

            return list(results)

    async def run_pattern_test(self) -> dict:
        """Run various patterns to identify session pool behavior."""
        all_results = {}

        print("\n" + "=" * 70)
        print("TEST 1: Rapid Fire (10 requests, no delay)")
        print("=" * 70)
        all_results["rapid_fire"] = await self.run_rapid_test(10, delay_ms=0)

        await asyncio.sleep(2)

        print("\n" + "=" * 70)
        print("TEST 2: With 500ms Delay (10 requests)")
        print("=" * 70)
        all_results["delayed_500ms"] = await self.run_rapid_test(10, delay_ms=500)

        await asyncio.sleep(2)

        print("\n" + "=" * 70)
        print("TEST 3: With 2000ms Delay (5 requests)")
        print("=" * 70)
        all_results["delayed_2000ms"] = await self.run_rapid_test(5, delay_ms=2000)

        await asyncio.sleep(2)

        print("\n" + "=" * 70)
        print("TEST 4: Parallel Requests (5 concurrent)")
        print("=" * 70)
        all_results["parallel"] = await self.run_parallel_test(5)

        await asyncio.sleep(2)

        print("\n" + "=" * 70)
        print("TEST 5: Random Jitter (10 requests, 100-1000ms random delay)")
        print("=" * 70)
        results = []
        async with httpx.AsyncClient(verify=False, timeout=60.0) as client:
            await self.authenticate(client)
            for i in range(10):
                result = await self.make_request(client, i + 1)
                results.append(result)
                status = "OK" if result.success else "FAIL"
                print(f"  [{i+1:2d}] {status} {result.elapsed_ms:4d}ms - {result.response_preview[:50]}")
                if i < 9:
                    jitter = random.uniform(0.1, 1.0)
                    await asyncio.sleep(jitter)
        all_results["random_jitter"] = results

        return all_results

    def analyze_results(self, all_results: dict) -> str:
        """Analyze test results and generate report."""
        report = []
        report.append("\n" + "=" * 70)
        report.append("SESSION POOL BEHAVIOR ANALYSIS")
        report.append("=" * 70)

        total_requests = 0
        total_failures = 0
        failure_patterns = {}

        for test_name, results in all_results.items():
            successes = sum(1 for r in results if r.success)
            failures = len(results) - successes
            total_requests += len(results)
            total_failures += failures

            report.append(f"\n{test_name.upper()}:")
            report.append(f"  Total: {len(results)}, Success: {successes}, Failed: {failures}")
            report.append(f"  Success Rate: {successes/len(results)*100:.1f}%")

            # Track failure patterns
            for r in results:
                if not r.success:
                    error_key = r.error_type or "Unknown"
                    failure_patterns[error_key] = failure_patterns.get(error_key, 0) + 1

            # Check for alternating pattern
            if len(results) >= 4:
                pattern = [r.success for r in results]
                alternating = all(pattern[i] != pattern[i+1] for i in range(len(pattern)-1))
                if alternating:
                    report.append("  [!] ALTERNATING PATTERN DETECTED!")

            # Check consecutive failures
            max_consecutive_fail = 0
            current_consecutive = 0
            for r in results:
                if not r.success:
                    current_consecutive += 1
                    max_consecutive_fail = max(max_consecutive_fail, current_consecutive)
                else:
                    current_consecutive = 0
            if max_consecutive_fail > 2:
                report.append(f"  [!] Max consecutive failures: {max_consecutive_fail}")

        report.append("\n" + "-" * 70)
        report.append("SUMMARY:")
        report.append(f"  Total Requests: {total_requests}")
        report.append(f"  Total Failures: {total_failures}")
        report.append(f"  Overall Success Rate: {(total_requests-total_failures)/total_requests*100:.1f}%")

        if failure_patterns:
            report.append("\n  Failure Types:")
            for error_type, count in sorted(failure_patterns.items(), key=lambda x: -x[1]):
                report.append(f"    - {error_type}: {count}")

        # Conclusions
        report.append("\n" + "-" * 70)
        report.append("CONCLUSIONS:")

        if total_failures == 0:
            report.append("  [OK] No failures detected - session pool appears healthy")
        elif total_failures / total_requests > 0.3:
            report.append("  [!] High failure rate (>30%) - likely session pool contamination")
            report.append("  [!] Consider using async endpoint or implementing retry logic")
        else:
            report.append("  [!] Intermittent failures detected")
            report.append("  [!] Pattern suggests session pool issues")

        if "UnexpectedWindow" in failure_patterns:
            report.append("  [!] 'Unexpected window' errors confirm dirty session pool")
            report.append("  [!] Previous operations left dialogs open in pooled sessions")

        return "\n".join(report)


async def main():
    print("=" * 70)
    print("P21 Transaction API - Session Pool Behavior Test")
    print("=" * 70)
    print(f"Server: {BASE_URL}")
    print(f"Time: {datetime.now().isoformat()}")

    tester = SessionPoolTester(BASE_URL, USERNAME, PASSWORD)

    # Run all pattern tests
    all_results = await tester.run_pattern_test()

    # Analyze and print report
    report = tester.analyze_results(all_results)
    print(report)

    # Save results to file
    output_file = Path(__file__).parent / "session_pool_results.json"
    results_json = {
        test_name: [
            {
                "attempt": r.attempt,
                "timestamp": r.timestamp,
                "elapsed_ms": r.elapsed_ms,
                "success": r.success,
                "status_code": r.status_code,
                "error_type": r.error_type,
                "error_message": r.error_message,
                "session_headers": r.session_headers
            }
            for r in results
        ]
        for test_name, results in all_results.items()
    }

    with open(output_file, "w") as f:
        json.dump(results_json, f, indent=2)

    print(f"\nDetailed results saved to: {output_file}")


if __name__ == "__main__":
    asyncio.run(main())

File: examples/csharp/Transaction/TestSessionPool.cs
// Transaction API - Session Pool Behavior Test
//
// Tests for session pool contamination issues.
// Mirrors: scripts/transaction/test_session_pool.py
//
// This script runs multiple test patterns to diagnose intermittent failures
// caused by dirty session pools:
//   1. Rapid fire requests (no delay)
//   2. Requests with 500ms delay
//   3. Requests with 2000ms delay
//   4. Parallel/concurrent requests
//   5. Random jitter delays
//
// Results help identify:
//   - Alternating success/failure patterns
//   - "Unexpected window" errors (dirty session)
//   - Failure rate vs request timing

using System.Diagnostics;
using System.Net.Http;
using System.Text;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using P21Examples.Common;

namespace P21Examples.Transaction;

/// <summary>
/// Tests P21 Transaction API session pool behavior with various
/// request patterns to detect contamination issues.
/// </summary>
public static class TestSessionPool
{
    /// <summary>
    /// Result of a single API call.
    /// </summary>
    private record TestResult(
        int Attempt,
        string Timestamp,
        int ElapsedMs,
        bool Success,
        int StatusCode,
        string? ErrorType = null,
        string? ErrorMessage = null,
        Dictionary<string, string>? SessionHeaders = null,
        string ResponsePreview = "");

    public static async Task RunAsync()
    {
        Console.WriteLine(new string('=', 70));
        Console.WriteLine("P21 Transaction API - Session Pool Behavior Test");
        Console.WriteLine(new string('=', 70));

        var config = P21Config.FromEnvironment();
        Console.WriteLine($"Server: {config.BaseUrl}");
        Console.WriteLine($"Time: {DateTime.Now:O}");

        // Create an HttpClient for raw control (no P21Client wrapper here,
        // because we need precise timing and header inspection)
        var handler = new HttpClientHandler();
        if (!config.VerifySsl)
        {
            handler.ServerCertificateCustomValidationCallback =
                HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
        }
        using var http = new HttpClient(handler) { Timeout = TimeSpan.FromSeconds(60) };

        // Authenticate
        var tokenResponse = await P21Auth.GetTokenAsync(http, config);
        P21Auth.SetAuthHeaders(http, tokenResponse.AccessToken);

        var uiServerUrl = await P21Auth.GetUiServerUrlAsync(http, config.BaseUrl);

        // Run all test patterns
        var allResults = new Dictionary<string, List<TestResult>>();

        // TEST 1: Rapid fire
        Console.WriteLine("\n" + new string('=', 70));
        Console.WriteLine("TEST 1: Rapid Fire (10 requests, no delay)");
        Console.WriteLine(new string('=', 70));
        allResults["rapid_fire"] = await RunRapidTestAsync(http, uiServerUrl, count: 10, delayMs: 0);

        await Task.Delay(2000);

        // TEST 2: 500ms delay
        Console.WriteLine("\n" + new string('=', 70));
        Console.WriteLine("TEST 2: With 500ms Delay (10 requests)");
        Console.WriteLine(new string('=', 70));
        allResults["delayed_500ms"] = await RunRapidTestAsync(http, uiServerUrl, count: 10, delayMs: 500);

        await Task.Delay(2000);

        // TEST 3: 2000ms delay
        Console.WriteLine("\n" + new string('=', 70));
        Console.WriteLine("TEST 3: With 2000ms Delay (5 requests)");
        Console.WriteLine(new string('=', 70));
        allResults["delayed_2000ms"] = await RunRapidTestAsync(http, uiServerUrl, count: 5, delayMs: 2000);

        await Task.Delay(2000);

        // TEST 4: Parallel requests
        Console.WriteLine("\n" + new string('=', 70));
        Console.WriteLine("TEST 4: Parallel Requests (5 concurrent)");
        Console.WriteLine(new string('=', 70));
        allResults["parallel"] = await RunParallelTestAsync(http, uiServerUrl, count: 5);

        await Task.Delay(2000);

        // TEST 5: Random jitter
        Console.WriteLine("\n" + new string('=', 70));
        Console.WriteLine("TEST 5: Random Jitter (10 requests, 100-1000ms random delay)");
        Console.WriteLine(new string('=', 70));
        allResults["random_jitter"] = await RunJitterTestAsync(http, uiServerUrl, count: 10);

        // Analyze and print report
        AnalyzeResults(allResults);

        // Save results to JSON file
        SaveResults(allResults);
    }

    /// <summary>
    /// Run sequential requests with a fixed delay between each.
    /// </summary>
    private static async Task<List<TestResult>> RunRapidTestAsync(
        HttpClient http, string uiServerUrl, int count, int delayMs)
    {
        var results = new List<TestResult>();

        for (var i = 0; i < count; i++)
        {
            var result = await MakeRequestAsync(http, uiServerUrl, i + 1);
            results.Add(result);

            var status = result.Success ? "OK" : "FAIL";
            Console.WriteLine($"  [{i + 1,2}] {status} {result.ElapsedMs,4}ms - {Truncate(result.ResponsePreview, 50)}");

            if (delayMs > 0 && i < count - 1)
            {
                await Task.Delay(delayMs);
            }
        }

        return results;
    }

    /// <summary>
    /// Run concurrent/parallel requests to stress test the session pool.
    /// </summary>
    private static async Task<List<TestResult>> RunParallelTestAsync(
        HttpClient http, string uiServerUrl, int count)
    {
        // Launch all requests concurrently
        var tasks = Enumerable.Range(1, count)
            .Select(i => MakeRequestAsync(http, uiServerUrl, i))
            .ToArray();

        var results = await Task.WhenAll(tasks);

        foreach (var result in results)
        {
            var status = result.Success ? "OK" : "FAIL";
            Console.WriteLine($"  [{result.Attempt,2}] {status} {result.ElapsedMs,4}ms - {Truncate(result.ResponsePreview, 50)}");
        }

        return results.ToList();
    }

    /// <summary>
    /// Run requests with random jitter between 100-1000ms.
    /// </summary>
    private static async Task<List<TestResult>> RunJitterTestAsync(
        HttpClient http, string uiServerUrl, int count)
    {
        var results = new List<TestResult>();
        var random = new Random();

        for (var i = 0; i < count; i++)
        {
            var result = await MakeRequestAsync(http, uiServerUrl, i + 1);
            results.Add(result);

            var status = result.Success ? "OK" : "FAIL";
            Console.WriteLine($"  [{i + 1,2}] {status} {result.ElapsedMs,4}ms - {Truncate(result.ResponsePreview, 50)}");

            if (i < count - 1)
            {
                var jitter = random.Next(100, 1001);
                await Task.Delay(jitter);
            }
        }

        return results;
    }

    /// <summary>
    /// Make a single Transaction API request and capture timing/headers.
    /// </summary>
    private static async Task<TestResult> MakeRequestAsync(
        HttpClient http, string uiServerUrl, int attempt)
    {
        var timestamp = DateTime.Now.ToString("O");
        var sw = Stopwatch.StartNew();

        var payload = BuildTestPayload();

        try
        {
            var content = new StringContent(
                JsonConvert.SerializeObject(payload),
                Encoding.UTF8,
                "application/json");

            var response = await http.PostAsync($"{uiServerUrl}/api/v2/transaction", content);

            sw.Stop();
            var elapsedMs = (int)sw.ElapsedMilliseconds;

            // Capture session-related response headers
            var sessionHeaders = new Dictionary<string, string>();
            foreach (var header in response.Headers)
            {
                var key = header.Key.ToLower();
                if (key.Contains("session") || key.Contains("x-p21") ||
                    key.Contains("server") || key.Contains("instance"))
                {
                    sessionHeaders[header.Key] = string.Join(", ", header.Value);
                }
            }

            if (response.IsSuccessStatusCode)
            {
                var body = await response.Content.ReadAsStringAsync();
                var data = JObject.Parse(body);
                var summary = data["Summary"] as JObject;
                var succeeded = summary?["Succeeded"]?.Value<int>() ?? 0;
                var failed = summary?["Failed"]?.Value<int>() ?? 0;
                var messages = data["Messages"] as JArray;

                if (succeeded > 0 && failed == 0)
                {
                    return new TestResult(
                        attempt, timestamp, elapsedMs,
                        Success: true, StatusCode: 200,
                        SessionHeaders: sessionHeaders,
                        ResponsePreview: $"Succeeded: {succeeded}");
                }
                else
                {
                    var errorMsg = messages?.FirstOrDefault()?.ToString() ?? "Unknown error";
                    return new TestResult(
                        attempt, timestamp, elapsedMs,
                        Success: false, StatusCode: 200,
                        ErrorType: "ValidationError",
                        ErrorMessage: Truncate(errorMsg, 200),
                        SessionHeaders: sessionHeaders,
                        ResponsePreview: $"Failed: {failed}, Messages: {messages?.Count ?? 0}");
                }
            }
            else
            {
                var errorText = await response.Content.ReadAsStringAsync();
                var errorType = errorText.Contains("Unexpected response window")
                    ? "UnexpectedWindow"
                    : "HTTPError";

                return new TestResult(
                    attempt, timestamp, elapsedMs,
                    Success: false, StatusCode: (int)response.StatusCode,
                    ErrorType: errorType,
                    ErrorMessage: Truncate(errorText, 200),
                    SessionHeaders: sessionHeaders,
                    ResponsePreview: Truncate(errorText, 100));
            }
        }
        catch (Exception ex)
        {
            sw.Stop();
            return new TestResult(
                attempt, timestamp, (int)sw.ElapsedMilliseconds,
                Success: false, StatusCode: 0,
                ErrorType: ex.GetType().Name,
                ErrorMessage: Truncate(ex.Message, 200));
        }
    }

    /// <summary>
    /// Build a SalesPricePage test payload that should succeed.
    /// </summary>
    private static JObject BuildTestPayload()
    {
        var timestamp = DateTime.Now.ToString("HHmmssfff");
        return new JObject
        {
            ["Name"] = "SalesPricePage",
            ["UseCodeValues"] = false,
            ["Transactions"] = new JArray
            {
                new JObject
                {
                    ["Status"] = "New",
                    ["DataElements"] = new JArray
                    {
                        new JObject
                        {
                            ["Name"] = "FORM.form",
                            ["Type"] = "Form",
                            ["Keys"] = new JArray(),
                            ["Rows"] = new JArray
                            {
                                new JObject
                                {
                                    ["Edits"] = new JArray
                                    {
                                        new JObject { ["Name"] = "price_page_type_cd", ["Value"] = "Supplier / Product Group" },
                                        new JObject { ["Name"] = "company_id", ["Value"] = "ACME" },
                                        new JObject { ["Name"] = "supplier_id", ["Value"] = 10.0 },
                                        new JObject { ["Name"] = "product_group_id", ["Value"] = "MISC" },
                                        new JObject { ["Name"] = "description", ["Value"] = $"SESSION-TEST-{timestamp}" },
                                        new JObject { ["Name"] = "pricing_method_cd", ["Value"] = "Source" },
                                        new JObject { ["Name"] = "source_price_cd", ["Value"] = "Supplier List Price" },
                                        new JObject { ["Name"] = "effective_date", ["Value"] = "2025-01-01" },
                                        new JObject { ["Name"] = "expiration_date", ["Value"] = "2030-12-31" },
                                        new JObject { ["Name"] = "totaling_method_cd", ["Value"] = "Item" },
                                        new JObject { ["Name"] = "totaling_basis_cd", ["Value"] = "Supplier List Price" },
                                        new JObject { ["Name"] = "row_status_flag", ["Value"] = "Active" }
                                    },
                                    ["RelativeDateEdits"] = new JArray()
                                }
                            }
                        },
                        new JObject
                        {
                            ["Name"] = "VALUES.values",
                            ["Type"] = "Form",
                            ["Keys"] = new JArray(),
                            ["Rows"] = new JArray
                            {
                                new JObject
                                {
                                    ["Edits"] = new JArray
                                    {
                                        new JObject { ["Name"] = "calculation_method_cd", ["Value"] = "Multiplier" },
                                        new JObject { ["Name"] = "calculation_value1", ["Value"] = "0.5" }
                                    },
                                    ["RelativeDateEdits"] = new JArray()
                                }
                            }
                        }
                    }
                }
            }
        };
    }

    /// <summary>
    /// Analyze all test results and print a summary report.
    /// </summary>
    private static void AnalyzeResults(Dictionary<string, List<TestResult>> allResults)
    {
        Console.WriteLine("\n" + new string('=', 70));
        Console.WriteLine("SESSION POOL BEHAVIOR ANALYSIS");
        Console.WriteLine(new string('=', 70));

        var totalRequests = 0;
        var totalFailures = 0;
        var failurePatterns = new Dictionary<string, int>();

        foreach (var (testName, results) in allResults)
        {
            var successes = results.Count(r => r.Success);
            var failures = results.Count - successes;
            totalRequests += results.Count;
            totalFailures += failures;

            Console.WriteLine($"\n{testName.ToUpper()}:");
            Console.WriteLine($"  Total: {results.Count}, Success: {successes}, Failed: {failures}");
            Console.WriteLine($"  Success Rate: {(double)successes / results.Count * 100:F1}%");

            // Track failure patterns
            foreach (var r in results.Where(r => !r.Success))
            {
                var errorKey = r.ErrorType ?? "Unknown";
                failurePatterns.TryGetValue(errorKey, out var count);
                failurePatterns[errorKey] = count + 1;
            }

            // Check for alternating success/failure pattern
            if (results.Count >= 4)
            {
                var pattern = results.Select(r => r.Success).ToList();
                var alternating = true;
                for (var i = 0; i < pattern.Count - 1; i++)
                {
                    if (pattern[i] == pattern[i + 1])
                    {
                        alternating = false;
                        break;
                    }
                }
                if (alternating)
                {
                    Console.WriteLine("  [!] ALTERNATING PATTERN DETECTED!");
                }
            }

            // Check consecutive failures
            var maxConsecutiveFail = 0;
            var currentConsecutive = 0;
            foreach (var r in results)
            {
                if (!r.Success)
                {
                    currentConsecutive++;
                    maxConsecutiveFail = Math.Max(maxConsecutiveFail, currentConsecutive);
                }
                else
                {
                    currentConsecutive = 0;
                }
            }
            if (maxConsecutiveFail > 2)
            {
                Console.WriteLine($"  [!] Max consecutive failures: {maxConsecutiveFail}");
            }
        }

        Console.WriteLine("\n" + new string('-', 70));
        Console.WriteLine("SUMMARY:");
        Console.WriteLine($"  Total Requests: {totalRequests}");
        Console.WriteLine($"  Total Failures: {totalFailures}");
        Console.WriteLine($"  Overall Success Rate: {(double)(totalRequests - totalFailures) / totalRequests * 100:F1}%");

        if (failurePatterns.Count > 0)
        {
            Console.WriteLine("\n  Failure Types:");
            foreach (var (errorType, count) in failurePatterns.OrderByDescending(x => x.Value))
            {
                Console.WriteLine($"    - {errorType}: {count}");
            }
        }

        // Conclusions
        Console.WriteLine("\n" + new string('-', 70));
        Console.WriteLine("CONCLUSIONS:");

        if (totalFailures == 0)
        {
            Console.WriteLine("  [OK] No failures detected - session pool appears healthy");
        }
        else if ((double)totalFailures / totalRequests > 0.3)
        {
            Console.WriteLine("  [!] High failure rate (>30%) - likely session pool contamination");
            Console.WriteLine("  [!] Consider using async endpoint or implementing retry logic");
        }
        else
        {
            Console.WriteLine("  [!] Intermittent failures detected");
            Console.WriteLine("  [!] Pattern suggests session pool issues");
        }

        if (failurePatterns.ContainsKey("UnexpectedWindow"))
        {
            Console.WriteLine("  [!] 'Unexpected window' errors confirm dirty session pool");
            Console.WriteLine("  [!] Previous operations left dialogs open in pooled sessions");
        }
    }

    /// <summary>
    /// Save detailed results to a JSON file.
    /// </summary>
    private static void SaveResults(Dictionary<string, List<TestResult>> allResults)
    {
        var outputFile = Path.Combine(AppContext.BaseDirectory, "session_pool_results.json");

        var jsonResults = new JObject();
        foreach (var (testName, results) in allResults)
        {
            var array = new JArray();
            foreach (var r in results)
            {
                array.Add(new JObject
                {
                    ["attempt"] = r.Attempt,
                    ["timestamp"] = r.Timestamp,
                    ["elapsed_ms"] = r.ElapsedMs,
                    ["success"] = r.Success,
                    ["status_code"] = r.StatusCode,
                    ["error_type"] = r.ErrorType,
                    ["error_message"] = r.ErrorMessage,
                    ["session_headers"] = r.SessionHeaders != null
                        ? JObject.FromObject(r.SessionHeaders)
                        : new JObject()
                });
            }
            jsonResults[testName] = array;
        }

        File.WriteAllText(outputFile, jsonResults.ToString(Formatting.Indented));
        Console.WriteLine($"\nDetailed results saved to: {outputFile}");
    }

    /// <summary>
    /// Truncate a string to a maximum length.
    /// </summary>
    private static string Truncate(string? value, int maxLength)
    {
        if (string.IsNullOrEmpty(value)) return "";
        return value.Length <= maxLength ? value : value[..maxLength];
    }
}
